src/dfu/dfu_main.cpp

```diff
@@ -0,0 +1,1002 @@
+ * @param cmd_name      Command Name
+ * @param cmd_id        Command ID
+ * @param num_values    Number of values the command writes
+ * @param values        Buffer storing the values to write
+ *
+ * @return              device control status
+ */
+control_ret_t command_set(Device * device, control_resid_t res_id, string cmd_name, control_cmd_t cmd_id, unsigned num_values, uint8_t * values)
+{
+    size_t data_len = num_values;
+    uint8_t * data = new uint8_t[data_len];
+
+    for (unsigned i = 0; i < num_values; i++)
+    {
+        memcpy(data + i, &values[i], 1);
+    }
+
+    control_ret_t ret = device->device_set(res_id, cmd_id, data, data_len);
+    int write_attempts = 1;
+
+    while(1)
+    {
+        if(write_attempts == 1000)
+        {
+            cerr << "Resource could not respond to the " << cmd_name << " write command."
+            << endl << "Check the audio loop is active." << endl;
+            exit(HOST_APP_ERROR);
+        }
+        if(ret == CONTROL_SUCCESS)
+        {
+            break;
+        }
+        else if(ret == SERVICER_COMMAND_RETRY)
+        {
+            ret = device->device_set(res_id, cmd_id, data, data_len);
+            write_attempts++;
+        }
+        else
+        {
+            check_cmd_error(cmd_name, "write", ret);
+        }
+    }
+
+    delete []data;
+    check_cmd_error(cmd_name, "write", ret);
+    return ret;
+}
+
+/** @brief List of supported CLI options */
+opt_t options[] = {
+    {"--help",                    "-h",        "display this information"                                                                                           },
+    {"--app-version",             "-av",       "print the version of this application",                                                                             },
+    {"--use",                     "-u",        "use specific hardware protocol, I2C, SPI and USB are available to use"                                              },
+    {"--verbose",                 "-vvv",      "enable debug prints"                                                                                                },
+    {"--upload-start",            "-us",       "set the first block transport number for the upload operation. Default is 0. Option valid only with upload commands"},
+    {"--version",                 "-v",        "read the version on the device",                                                                                    },
+    {"--download",                "-d",        "download upgrade image stored in the specified path"                                                                },
+    {"--upload-factory",          "-uf",       "upload factory image and save it in the specified path."                                                            },
+    {"--upload-upgrade",          "-uu",       "upload upgrade image and save it in the specified path."                                                            },
```

Comment:
These two options have a dot at the end of a help string. I would suggest removing them as none of the help strings are capitalised and end with a dot.

Reply:
Good catch!